### PR TITLE
Set up foxtrot user infrastructure

### PR DESCRIPTION
**Pull request by juliaogris**
Create core foxtrot `User` data structure, Database layer to create and
update users and `Authenticator` validating password on login and creating
password hash with bcrypt on user registration.

Generate `schema.go` from `sql/schema.go` so that sql commands can be used
directly and the same commands can be used in go without packing
shenanigans.

Misc: Update schema insignificantly, set up Makefile and .golangci.yaml linter config.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+# --- Build --------------------------------------------------------------------
+
+build: | $(O)  ## Build binaries of directories in ./cmd to out/
+	go build -o $(O) ./cmd/...
+
+install:  ## Build and install binaries in $GOBIN or $GOPATH/bin
+	go install ./cmd/...
+
+run: build ## Run foxtrot server
```
Comment (camh-) [inline Makefile]: nit: two spaces before `##`

File: .golangci.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    - staticcheck
+    - structcheck
+    - stylecheck
+    - typecheck
+    - unconvert
+    - unparam
+    - unused
+    - varcheck
+    - whitespace
+    - errorlint
```
Comment (camh-) [inline .golangci.yaml]: These stop being alphabetical about here.

File: .golangci.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    - goerr113
+    - lll
```
Comment (camh-) [inline .golangci.yaml]: goerr113 and lll are also listed in the "disabled" comment.

File: Makefile [hunk 4]
```diff
@@ -22,6 +22,9 @@ install:  ## Build and install binaries in $GOBIN or $GOPATH/bin
 run: build ## Run foxtrot server
 	$(O)/foxtrot
 
+pkg/foxtrot/schema.go: sql/schema.sql
+	@printf '//generated DO NOT EDIT\n\npackage foxtrot\n\nconst schema = `%s`\n' "$$(cat $<)" > $@
```
Comment (camh-) [inline Makefile:26]: note this only works if `sql/schema.sql` has no backticks in it. but i think this is good enough until we start embedding it with the new go embedding.

File: pkg/foxtrot/db.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+func newDB(dsnURI string) (*db, error) {
+	conn, err := sql.Open("sqlite3", dsnURI)
+	if err != nil {
+		return nil, errs.New(errDBInitialisation, err)
+	}
+
+	db := &db{conn: conn}
+	if err := db.setupSchema(); err != nil {
+		_ = db.conn.Close()
```
Comment (camh-) [inline pkg/foxtrot/db.go]: i'd use `db.close()` here i think

File: pkg/foxtrot/db.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+func (db *db) setupSchema() error {
+	selectVersionStr := "SELECT version FROM schema"
+	version := ""
+	err := db.conn.QueryRow(selectVersionStr).Scan(&version)
+	expectedVersion := "v0.0.1"
+	if err == nil && version == expectedVersion {
+		return nil
+	} else if err == nil {
```
Comment (camh-) [inline pkg/foxtrot/db.go]: perhaps swap the two branches of the `if`:
```go
if err == nil && version != expectedVersion {
    return errs...
} else if err == nil {
    return nil
}

// Assume an error means the schema has not been created, so create it now.
```

I think that makes the error condition a little clearer. Also that comment makes it a little easier to understand what is going on I think.

File: pkg/foxtrot/db.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		return errs.Errorf("%v: bad version '%s' expected '%s'", errDBInitialisation, version, expectedVersion)
+	}
+	_, err = db.conn.Exec(schema)
+	if err != nil {
```
Comment (camh-) [inline pkg/foxtrot/db.go]: combine if with previous line

File: pkg/foxtrot/db_test.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+func TestCreateGetUser(t *testing.T) {
+	db, err := newDB("")
+	require.NoError(t, err)
+	defer db.close()
+
+	u := &User{Name: "alice", passwordHash: "###"}
+	err = db.createUser(context.Background(), u)
+	require.NoError(t, err)
+
+	u2, err := db.getUser(context.Background(), "alice")
```
Comment (camh-) [inline pkg/foxtrot/db_test.go:23]: where the vars have no name :musical_note: (u2, get it?)

File: pkg/foxtrot/db_test.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	require.NoError(t, err)
+	require.Equal(t, u, u2)
+}
+
+func TestGetUserErr(t *testing.T) {
+	db, err := newDB("")
+	require.NoError(t, err)
+	defer db.close()
+
+	_, err = db.getUser(context.Background(), "MISSING")
```
Comment (camh-) [inline pkg/foxtrot/db_test.go:33]: Miss Ing will be disappointed to know she is an error.

File: pkg/foxtrot/foxtrot.go [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+// User is a core foxtrot data structure representing a user entry in
+// the database as well as the data related to a user as presented in
+// the web UI.
+type User struct {
+	Name      string `json:"name"`
+	AvatarURL string `json:"avatarURL,omitempty"`
+
+	passwordHash string
+	avatarBlob   []byte
```
Comment (camh-) [inline pkg/foxtrot/foxtrot.go]: not sure this needs "Blob" on the end of the name,

File: pkg/foxtrot/auth_test.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+func TestRegisterLogin(t *testing.T) {
+	db, err := newDB("")
+	require.NoError(t, err)
+	defer db.close()
+
+	a := authenticator{db: db}
+	u := &User{Name: "Alice"}
+	err = a.register(context.Background(), u, "Pa$$w0rd")
```
Comment (camh-) [inline pkg/foxtrot/auth_test.go:18]: i would appreciate it if you would not add my password to the repository for all to see.

File: pkg/foxtrot/auth.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+var (
+	errAuth         = errors.New("user authentication error")
+	errPasswordHash = errors.New("password hash creation err")
+)
+
+type authenticator struct {
+	db *db
+}
+
+func (a *authenticator) register(ctx context.Context, u *User, password string) error {
```
Comment (camh-) [inline pkg/foxtrot/auth.go:20]: my feeling is that the user parameter should be a string. registration inputs will be username and password. a User struct would not exist at this point, yes?

File: .golangci.yaml (hunk 2 above)
Comment (juliaogris) [reply .golangci.yaml]: sorry, fixed.

File: Makefile (hunk 1 above)
Comment (juliaogris) [reply Makefile]: fixed, sorry.

File: .golangci.yaml (hunk 3 above)
Comment (juliaogris) [reply .golangci.yaml]: sorry, fixed.

File: pkg/foxtrot/db.go (hunk 5 above)
Comment (juliaogris) [reply pkg/foxtrot/db.go]: yes! I initially had prepared statements which I closed in db.close, which caused a panic (even with not nil guard).....
this was a temporary fix. 
done.

File: pkg/foxtrot/db.go (hunk 6 above)
Comment (juliaogris) [reply pkg/foxtrot/db.go]: YES! thanks.

File: pkg/foxtrot/db.go (hunk 7 above)
Comment (juliaogris) [reply pkg/foxtrot/db.go]: done.

File: pkg/foxtrot/db_test.go (hunk 8 above)
Comment (juliaogris) [reply pkg/foxtrot/db_test.go:23]: totally. I love [the streets](https://music.youtube.com/watch?v=WpfoHIoii6A&feature=share).

File: pkg/foxtrot/db_test.go (hunk 9 above)
Comment (juliaogris) [reply pkg/foxtrot/db_test.go:33]: I'm Miss Ing. Dipl-Ing is my legal title for having finished my CS degree (Diplomingeneur).

File: pkg/foxtrot/foxtrot.go (hunk 10 above)
Comment (juliaogris) [reply pkg/foxtrot/foxtrot.go]: ok.

File: pkg/foxtrot/auth.go (hunk 12 above)
Comment (juliaogris) [reply pkg/foxtrot/auth.go:20]: not sure. would you in a signup flow do all: username, nickname, fullname, avatar, password, phone number, address, relationship status?

Comment (juliaogris) [reply pkg/foxtrot/auth.go:20]: if yes, then i think it is ok to expect a user struct as input.

File: pkg/foxtrot/auth_test.go (hunk 11 above)
Comment (juliaogris) [reply pkg/foxtrot/auth_test.go:18]: 👅

File: pkg/foxtrot/db_test.go (hunk 8 above)
Comment (camh-) [reply pkg/foxtrot/db_test.go:23]: I will listen to that later

File: pkg/foxtrot/auth.go (hunk 12 above)
Comment (camh-) [reply pkg/foxtrot/auth.go:20]: relationship status? what are we building here? ;-)

I'm not fussed, but my uninformed expectation is that those things would not be provided on registration.

**Review by camh- — APPROVED** [review]
LGTM :brazil: I've left a couple of comments unresolved where I have replied. Feel free to resolve.

**Pull request merged**